src/classes/CW_AudienceGroupCompControlerTest.cls

```diff
@@ -0,0 +1,270 @@
+			//Chech Error Messages
+			system.assertEquals(true  ,ApexPages.hasMessages());
+			ApexPages.getMessages().clear();
+			compCtrl.newRec.Name='Test_Name2';
+			compCtrl.newRec.Is_Control_Group__c = false;
+			compCtrl.newRec.Control_Group__c=listGroups[1].Id;
+			compCtrl.Cancel();
+			
+			compCtrl.selectAll = true;
+			compCtrl.selectDeselectAllItems();
+			system.assertEquals(5     ,compCtrl.getSelectedCount());
+			system.assertEquals(false ,compCtrl.getDeleteDisabled());
+			
+			compCtrl.selectAll = false;
+			compCtrl.selectDeselectAllItems();
+			system.assertEquals(0     ,compCtrl.getSelectedCount());
+			system.assertEquals(true  ,compCtrl.getDeleteDisabled());
+			
+			compCtrl.selectedGroupIds.Add(listGroups[3].Id);
+			compCtrl.selectedGroupIds.Add(listGroups[4].Id);
+			system.assertEquals(false  ,compCtrl.getDeleteDisabled());
+			compCtrl.getDeleteMessage();
+			compCtrl.deleteGroups();
+			system.assertEquals(14    , RecList.Size());
+			system.assertEquals(5     ,compCtrl.campGroups.Size());
+			system.assertEquals(true  ,compCtrl.getOutputMode());
+			//Tray to delete used control groups
+			compCtrl.selectedGroupIds.Add(listGroups[0].Id);
+			compCtrl.selectedGroupIds.Add(listGroups[1].Id);
+			compCtrl.selectAll = false;
+			compCtrl.selectDeselectAllItems();
+			
+			compCtrl.getDeleteMessage();
+			compCtrl.deleteGroups();
+			
+			//check for errors
+			system.assertEquals(true  ,ApexPages.hasMessages());
+			system.assertEquals(14    , RecList.Size());
+			
+			compCtrl.contextItem = listGroups[4].Id;
+			compCtrl.doSelectItem();
+			system.assertEquals(1     ,compCtrl.getSelectedCount());
+			compCtrl.doDeselectItem();
+			system.assertEquals(0     ,compCtrl.getSelectedCount());
+			
+			compCtrl.last();
+			system.assertEquals(4     ,compCtrl.campGroups.Size());
+			
+			compCtrl.first();
+			system.assertEquals(5     ,compCtrl.campGroups.Size());
+			
+			compCtrl.doNext();
+			system.assertEquals(2     ,compCtrl.getPageNumber());
+			system.assertEquals(true  ,compCtrl.getHasPrevious());
+			system.assertEquals(true  ,compCtrl.getHasNext());
+
+			
+			compCtrl.doPrevious();
+			system.assertEquals(false  ,compCtrl.getHasPrevious());
+			system.assertEquals(true   ,compCtrl.getHasNext());
+
+			compCtrl.editGroups();
+			system.assertEquals(false ,compCtrl.getOutputMode());
+			system.assertEquals(false ,compCtrl.savedByParent);
+			compCtrl.savedByParent =true;
+			system.assertEquals(true  ,compCtrl.getOutputMode());
+			compCtrl.Cancel();
+			system.assertEquals(true  ,compCtrl.getOutputMode());
+	    }
+	}
+	static testMethod void WraperClasssControllerTest() 
+	{
+        initCustomSettings();
+        
+        CW_Campaign__c camp = TestDataUtil.createCampaign();
+        insert camp;
+        
+        List<CW_Audience_Group__c> listGroups = TestDataUtil.createAudienceGroups(camp.Id,5);
+        insert listGroups;
+        
+        User thisUser = [SELECT Id FROM User WHERE Id = :UserInfo.getUserId()];
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.